### PR TITLE
feat: add torchdynamo notebook

### LINKED ISSUE #107: Pytorch 1.12 breaks GPT-2 (and probably T5)
Updating Pytorch to version 1.12 breaks GPT-2 unit test.
The issue is related to `TensorRT`:

```log
[06/29/2022-20:09:34] [TRT] [W] ShapedWeights.cpp:173: Weights transformer.h.1.attn.c_attn.weight has been transposed with permutation of (1, 0)! If you plan on overwriting the weights with the Refitter API, the new weights must be pre-transposed.
[06/29/2022-20:09:34] [TRT] [W] onnx2trt_utils.cpp:392: One or more weights outside the range of INT32 was clamped
[06/29/2022-20:09:34] [TRT] [W] building engine. depending on model size this may take a while
[06/29/2022-20:09:34] [TRT] [E] 4: [network.cpp::validate::2633] Error Code 4: Internal Error (Network must have at least one output)
[06/29/2022-20:09:34] [TRT] [E] 2: [builder.cpp::buildSerializedNetwork::609] Error Code 2: Internal Error (Assertion enginePtr != nullptr failed. )
```

related to #102

### LINKED ISSUE #112: simple benchmark with torchdynamo
`Torchdynamo` is a prototype on which Pytorch team is working right now.
It supposed to offer the flexibility of Python and the optimization level of computation graph compilation.
It supports many backends, including `tvm`, `tensorrt` and `nvfuser`.

Static graph compilation is not always easy, we may want to perform a rapid benchmark to compare at least classic approach (`onnx runtime` or `tensorrt`) with dynamo and the new `nvFuser` (unlike precedent default fusers, NNC, it supports dynamic shape).

Some links:

* project: https://github.com/pytorch/torchdynamo
* HF benchmark: https://github.com/pytorch/torchdynamo/blob/main/benchmarks/huggingface.py
* `NNC` presentation: https://dev-discuss.pytorch.org/t/nnc-walkthrough-how-pytorch-ops-get-fused/125
* `nvFuser`: https://www.nvidia.com/en-us/on-demand/session/gtcspring21-s31952/

To install `Pytorch-TensorRT` for the `TensorRT` backend:

```bash
pip3 install git+https://github.com/pytorch/torchdynamo.git@ee1b62f4947a94a87c288f38711de76bfaa2ffd4#egg=torchdynamo
git clone https://github.com/pytorch/TensorRT.git
cd TensorRT/py && python setup.py install --fx-only && cd ..
```

### PR DESCRIPTION
**Pull request by pommedeterresautee**
* add torchdynamo notebook. fix #112
* fix TensorRT bug (add a new constant folding step). fix #107 
* simplify dependencies (remove version pinning not needed anymore)
* align output type of trt inference function to other inference function (from tuple of tensors to dict[tensor name, tensor value])

### CODE REVIEW

File: demo/torchdynamo/dynamo_utils.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+    gc.collect()  # delete all deletable objects so CUDA memory can be freed by empty_cache
+    torch.cuda.empty_cache()
+    timings: List[List[float]] = list()
+    for is_warmup in [True, True, False]:
+        for batch_size, seq_lens in shapes.items():
+            batch_timings: List[float] = list()
+            for seq_len in seq_lens:
+                inputs = get_pytorch_input(size=(batch_size, seq_len))
+                latencies = list()
+                nb_retry = 1 if is_warmup else 5
```
Comment (gaetansnl) [inline demo/torchdynamo/dynamo_utils.py:67]: IMO it's better to execute the exact same code in warmup and simply get the last result

**Review by gaetansnl — COMMENTED** [review]
Tested in Docker
I could not test the end with tensorrt (I think there is a problem with version described in the comments). 
I get a lot less noise in the graphs on my side. It may change graph interpretation.

File: demo/torchdynamo/dynamo_utils.py [hunk 2]
```diff
@@ -0,0 +1,193 @@
+#  Copyright 2022, Lefebvre Dalloz Services
```
[the remark this answers is not in the document]
Comment (pommedeterresautee) [reply demo/torchdynamo/dynamo_utils.py:1]: tensorrt -> it depends of the version installed in the docker image uses, I use 8.4 and it works
torchdynamo installation fails if done with the other dependencies, not sure why

File: requirements_gpu.txt [hunk 3]
```diff
@@ -1,6 +1,5 @@
 onnxruntime-gpu
-torch==1.11.0+cu113
-nvidia-tensorrt==8.4.1.5
+nvidia-tensorrt
```
[the remark this answers is not in the document]
Comment (pommedeterresautee) [reply requirements_gpu.txt:2]: for torch, it s now moved to the main requirement file, as there is no more need to precise the version, just by providing the custom repo for cuda 11.X it installs the right version.
unpinned nvidia trt as most deps an unpinned anyway.

File: demo/torchdynamo/dynamo_utils.py (hunk 1 above)
Comment (pommedeterresautee) [reply demo/torchdynamo/dynamo_utils.py:67]: ok

File: demo/torchdynamo/dynamo_utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (pommedeterresautee) [reply demo/torchdynamo/dynamo_utils.py:1]: By rerunning the notebook and shutting done any background process which may have an impact I got similar shapes (but not same speedups as we have not the same GPU)

**Review by gaetansnl — APPROVED** [review]

**Pull request merged**